### PR TITLE
Enhancements in Remote Debug Dialog

### LINKED ISSUE #1994: Make connecting to localhost gdbserver easier
**Is your feature request related to a problem? Please describe.**

When connecting to gdbserver multiple times in a row entering data each time is a bit annoying

**Describe the solution you'd like**

* Accept empty IP address field as localhost

**Describe alternatives you've considered**

* Remember IP address and port from previous run

**Additional context**

### PR DESCRIPTION
**Pull request by nk521**
**Detailed description**

Adds history listing feature in Remote Debug Dialog with some minor enhancements.
* Placeholders
* Remembers the previous connected paths/ips
* Blank IP will be treated as localhost

**Test plan (required)**

Placeholders and history box with previous connected IPs/Paths
image

Just click on any one and it'll automatically fill the details in dialog box
image

**Closing issues**
closes #1994

### CODE REVIEW

**Comment by XVilka** [comment]
Is it possible to separate the items from the history list visually? Now it looks a bit weird.

**Comment by nk521** [comment]
Sure. Will make a patch after sometime :)

**Comment by nk521** [comment]
Hey @XVilka, i'm not quite sure what you mean by `items from history list`. Previously i interpreted it as the space between values in history list. It looks something like this now (less space between list values). Can you confirm if this is what you meant? If not, let me know.

image

Also the right click menu now works.
image

**Comment by ITAYC0HEN** [comment]
I think he meant a line separator. But it might be the user theme that causes it to look like this.
Anyways he probably imagines something like this:

before:
```
1
2
3
```

after:
```
1
________
2
________
3
________
```

File: src/dialogs/RemoteDebugDialog.cpp [hunk 1]
```diff
@@ -155,6 +163,76 @@ void RemoteDebugDialog::setDebugger(QString debugger)
...
+    ui->recentsIpListWidget->takeItem(ui->recentsIpListWidget->currentRow());
+
+    ui->ipEdit->clear();
+
+}
+
+void RemoteDebugDialog::on_actionClear_all_triggered()
+{
+    QSettings settings;
+    QStringList ips = settings.value("recentIpList").toStringList();
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: Why do you need to fetch the list? you can simply  use empty `QStringList` when you do `settings.setValue("recentIpsList", ...`

File: src/dialogs/RemoteDebugDialog.ui [hunk 2]
```diff
@@ -72,19 +58,40 @@
        <bool>false</bool>
       </property>
       <property name="placeholderText">
-       <string notr="true"/>
+       <string notr="true">Enter IP Address</string>
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.ui]: Set "127.0.0.1" as default text + placeholder

File: src/dialogs/RemoteDebugDialog.ui [hunk 3]
```diff
@@ -130,7 +133,7 @@
        <bool>false</bool>
       </property>
       <property name="placeholderText">
-       <string notr="true"/>
+       <string notr="true">Enter Port</string>
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.ui]: Set the debugger's default port (check online fore the different options)

File: src/dialogs/RemoteDebugDialog.cpp [hunk 4]
```diff
@@ -155,6 +163,76 @@ void RemoteDebugDialog::setDebugger(QString debugger)
...
+        const QString ip = it.next();
+        const QString text = QString("%1").arg(ip);
+        QListWidgetItem *item = new QListWidgetItem(
+            text
+        );
+        item->setData(Qt::UserRole, ip);
+        ui->recentsIpListWidget->addItem(item);
+    }
+
+    settings.setValue("recentIpList", ips);
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: Why do you need to ser it back? You didn't change this list.

File: src/dialogs/RemoteDebugDialog.cpp [hunk 5]
```diff
@@ -155,6 +163,76 @@ void RemoteDebugDialog::setDebugger(QString debugger)
...
+    return !ips.isEmpty();
+}
+
+void RemoteDebugDialog::on_recentsIpListWidget_itemClicked(QListWidgetItem *item)
+{
+    QVariant data = item->data(Qt::UserRole);
+    QString ipport = data.toString();
+    if (!ipport.indexOf(GDB_URI_PREFIX)) {
+        ui->debuggerCombo->setCurrentIndex(ui->debuggerCombo->findText(GDBSERVER));
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: you can simply use `setCurrentText` here :)

File: src/dialogs/RemoteDebugDialog.cpp [hunk 6]
```diff
@@ -155,6 +163,76 @@ void RemoteDebugDialog::setDebugger(QString debugger)
...
+        int last_colon = ipport.lastIndexOf(QString(":"));
+        QString port_temp = ipport.mid(last_colon + 1, ipport.length());
+        QString ip_temp = ipport.mid(6, ipport.length() - port_temp.length() - 7);
+        ui->ipEdit->setText(ip_temp);
+        ui->portEdit->setText(port_temp);
+    } else if (!ipport.indexOf(WINDBG_URI_PREFIX)) {
+        ui->debuggerCombo->setCurrentIndex(ui->debuggerCombo->findText(WINDBGPIPE));
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: you can simply use setCurrentText here :)

File: src/dialogs/RemoteDebugDialog.cpp [hunk 7]
```diff
@@ -155,6 +163,76 @@ void RemoteDebugDialog::setDebugger(QString debugger)
...
+
+void RemoteDebugDialog::on_recentsIpListWidget_itemClicked(QListWidgetItem *item)
+{
+    QVariant data = item->data(Qt::UserRole);
+    QString ipport = data.toString();
+    if (!ipport.indexOf(GDB_URI_PREFIX)) {
+        ui->debuggerCombo->setCurrentIndex(ui->debuggerCombo->findText(GDBSERVER));
+        int last_colon = ipport.lastIndexOf(QString(":"));
+        QString port_temp = ipport.mid(last_colon + 1, ipport.length());
+        QString ip_temp = ipport.mid(6, ipport.length() - port_temp.length() - 7);
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: why "_temp"?

Comment (ITAYC0HEN) [reply src/dialogs/RemoteDebugDialog.cpp]: Try not to use hardcoded (6, -7)

File: src/dialogs/RemoteDebugDialog.cpp [hunk 8]
```diff
@@ -169,7 +247,12 @@ QString RemoteDebugDialog::getUri() const
 
 QString RemoteDebugDialog::getIp() const
 {
-    return ui->ipEdit->text();
+    auto temp = ui->ipEdit->text();
+    if (temp == "" || temp == "localhost") {
+        return "127.0.0.1";
+    } else {
+        return temp;
+    }
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: why is this needed? The protocols should handle URI and not only IP, please make sure

File: src/dialogs/RemoteDebugDialog.cpp [hunk 9]
```diff
@@ -155,6 +163,76 @@ void RemoteDebugDialog::setDebugger(QString debugger)
     ui->debuggerCombo->setCurrentIndex(ui->debuggerCombo->findText(debugger));
 }
 
+void RemoteDebugDialog::on_actionRemove_item_triggered()
```
Comment (karliss) [inline src/dialogs/RemoteDebugDialog.cpp]: Please don't use name based signal autoconnect. Use the connect as instructed in [developer documentaton](https://cutter.re/docs/developers-docs.html#connecting-signals) . Name based one can easily and silently break during refactoring.

File: src/dialogs/RemoteDebugDialog.cpp (hunk 8 above)
Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: I added this because karliss described it as the solution they want. (#1994)
> Accept empty IP address field as localhost.

File: src/dialogs/RemoteDebugDialog.cpp (hunk 7 above)
Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: "_temp" because I was struggling to name my variables good :P 

I was looking for a workaround of not using hardcoded values but I failed to find any. I'll give it another try :)

File: src/dialogs/RemoteDebugDialog.ui [hunk 10]
```diff
@@ -142,10 +145,42 @@
...
+      </property>
+      <property name="lineWidth">
+       <number>0</number>
+      </property>
+      <property name="resizeMode">
+       <enum>QListView::Adjust</enum>
+      </property>
+      <property name="spacing">
+       <number>2</number>
+      </property>
```
Comment (karliss) [inline src/dialogs/RemoteDebugDialog.ui]: Don't make unnecessary style modifications for single widget. If you don't like how a list widget in current theme looks -> modify or create a new theme. At this point easiest way of removing the modifications is to remove them by editing .ui file in text editor.

Comment (nk521) [reply src/dialogs/RemoteDebugDialog.ui]: I modified them so they match that newfile dialog screen. Anyways, I'll remove these lines.

File: src/dialogs/RemoteDebugDialog.cpp (hunk 8 above)
Comment (karliss) [reply src/dialogs/RemoteDebugDialog.cpp]: I didn't say anything about handling of string "localhost". Name lookup would be better done as separate issue, the code before and after the PR doesn't support it.  I don't like having special handling for localhost as it gives false impression that the field supports more than IP address, which it currently doesn't.

Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: Then I should remove "localhost" but keep empty string check as it is. Right?

Comment (ITAYC0HEN) [reply src/dialogs/RemoteDebugDialog.cpp]: If the current impl does not support URI and only IP, then alright, use only IP. I absolutely agree with Karliss, no need for "localhost".

That said, my suggestion for the implementation is.

0. Placeholder: 127.0.0.1
1. If no history exists, then set the value to 127.0.0.1
2. If user has usage history, then use the last item from the history.

File: src/dialogs/RemoteDebugDialog.ui (hunk 2 above)
Comment (nk521) [reply src/dialogs/RemoteDebugDialog.ui]: done

File: src/dialogs/RemoteDebugDialog.ui (hunk 3 above)
Comment (nk521) [reply src/dialogs/RemoteDebugDialog.ui]: done

File: src/dialogs/RemoteDebugDialog.cpp (hunk 4 above)
Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: Ah my bad. Anyways, made changes.

File: src/dialogs/RemoteDebugDialog.cpp (hunk 5 above)
Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: I'll avoid such small mistakes in future .-. 
done

File: src/dialogs/RemoteDebugDialog.cpp (hunk 9 above)
Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: Done. Will push changes soon

File: src/dialogs/RemoteDebugDialog.cpp [hunk 11]
```diff
@@ -173,32 +182,63 @@ void RemoteDebugDialog::on_actionRemove_item_triggered()
...
 
+/*
+* @brief Fills the remote debug dialog form with given string
+* Eg: gdb://127.0.0.1:8754 or windbg:///tmp/abcd
+*/
+void RemoteDebugDialog::fillFormData(QString formdata)
+{
+    // the PREFIX should be in start
+    // indexOf should return 0 if thats the case
+    if (!formdata.indexOf(GDB_URI_PREFIX)) {
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: Better use [QString::startsWith](https://doc.qt.io/qt-5/qstring.html#startsWith)

File: src/dialogs/RemoteDebugDialog.cpp [hunk 12]
```diff
@@ -173,32 +182,63 @@ void RemoteDebugDialog::on_actionRemove_item_triggered()
...
+        ui->debuggerCombo->setCurrentText(GDBSERVER);
+        int last_colon = formdata.lastIndexOf(QString(":"));
+        QString port_temp = formdata.mid(last_colon + 1, formdata.length());
+        // length of "gdb://" is 6
+        // TODO: Remove the hardcoded values
+        QString ip_temp = formdata.mid(6, formdata.length() - port_temp.length() - 7);
+        ui->ipEdit->setText(ip_temp);
+        ui->portEdit->setText(port_temp);
+    } else if (!formdata.indexOf(WINDBG_URI_PREFIX)) {
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: Better use [QString::startsWith](https://doc.qt.io/qt-5/qstring.html#startsWith)

File: src/dialogs/RemoteDebugDialog.cpp [hunk 13]
```diff
@@ -173,32 +182,63 @@ void RemoteDebugDialog::on_actionRemove_item_triggered()
...
+void RemoteDebugDialog::fillFormData(QString formdata)
+{
+    // the PREFIX should be in start
+    // indexOf should return 0 if thats the case
+    if (!formdata.indexOf(GDB_URI_PREFIX)) {
+        ui->debuggerCombo->setCurrentText(GDBSERVER);
+        int last_colon = formdata.lastIndexOf(QString(":"));
+        QString port_temp = formdata.mid(last_colon + 1, formdata.length());
+        // length of "gdb://" is 6
+        // TODO: Remove the hardcoded values
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: what about `GDB_URI_PREFIX.length()`?

File: src/dialogs/RemoteDebugDialog.cpp [hunk 14]
```diff
@@ -173,32 +182,63 @@ void RemoteDebugDialog::on_actionRemove_item_triggered()
...
+        QString ip_temp = formdata.mid(6, formdata.length() - port_temp.length() - 7);
+        ui->ipEdit->setText(ip_temp);
+        ui->portEdit->setText(port_temp);
+    } else if (!formdata.indexOf(WINDBG_URI_PREFIX)) {
+        ui->debuggerCombo->setCurrentText(WINDBGPIPE);
+        // length of "windbg://" is 9
+        QString path_temp = formdata.mid(9, formdata.length());
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: Same, what about `WINDBG_URI_PREFIX.length()`?

File: src/dialogs/RemoteDebugDialog.cpp [hunk 15]
```diff
@@ -208,29 +248,31 @@ bool RemoteDebugDialog::fillRecentIpList()
...
+        // fill recentsIpListWidget
         ui->recentsIpListWidget->addItem(item);
     }
 
-    settings.setValue("recentIpList", ips);
+    if (ips.isEmpty())
+    {
+        ui->ipEdit->setText("127.0.0.1");
+        ui->portEdit->clear();
+    } else {
```
Comment (ITAYC0HEN) [inline src/dialogs/RemoteDebugDialog.cpp]: You already put the default value of `ipEdit` to 127.0.0.1 on the UI file, So why is this necessary?

Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: I never added the default value of `ipEdit` in UI file. This is why this piece of code is here.

Comment (ITAYC0HEN) [reply src/dialogs/RemoteDebugDialog.cpp]: So don't make it only a placeholder, but also the value :)
https://github.com/radareorg/cutter/blob/619223cd192ec4c787b7a0f9bbcb10a8917543c6/src/dialogs/RemoteDebugDialog.ui#L60-L62

Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: done.

File: src/dialogs/RemoteDebugDialog.cpp (hunk 14 above)
Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: done .-.

File: src/dialogs/RemoteDebugDialog.cpp (hunk 13 above)
Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: done.

Comment (nk521) [reply src/dialogs/RemoteDebugDialog.cpp]: had to cast it to QString before, btw

**Comment by nk521** [comment]
@ITAYC0HEN I'll push the user-documentation part in some time :)

**Comment by nk521** [comment]
Added line separators
image

**Comment by karliss** [comment]
As for the line separators I already said my opinion - don't touch the default theme.

I am guessing that the comment by XVilka was due to having unnecessary style modifications copied from initial page. Just using default look of list widget should be fine.

File: src/dialogs/RemoteDebugDialog.cpp (hunk 9 above)
Comment (karliss) [reply src/dialogs/RemoteDebugDialog.cpp]: Use camel case for method names instead of snake_case like the rest of the code.

**Comment by ITAYC0HEN** [comment]
I agree with karliss, fon't touch this :)

**Comment by karliss** [comment]
Before making a comment I recommended going through all the changes and asking yourself if you wanted to make them. It helps catching accidental changes and unfinished or testing code you didn't want to commit. New files are usually obvious but unintentional source modifications in the same file you made intentional changes are harder to notice after the fact.

There two approaches to this. One is to use `git diff --staged` or it's equivalent in your git gui. Other is to carefully stage changes one by instead of staging everything.

.gitignore is still useful for avoiding common mistakes and reducing noise while staging.

**Comment by ITAYC0HEN** [comment]
I have more suggestions:

1. make sure port is valid, allow only digits, and maximum value (of highest port)
2. Don't show the previous connection view if no preview exists, it's confusing. 
3. Usually, such widgets are using combo boxes for previous values. I think the history view you created is very nice and I suggest to give it extra spices. I suggest to filter its content based on the text the user types in either IP or Port.  Check the theme color filtering in the theme editor for example. This should be possible and would make it veyr nice. It will be like a quick filter

**Comment by XVilka** [comment]
Due to the recent rename of `windbg://` into `winkd://` and some corresponding refactoring this *might* need an update.

**Comment by xarkes** [comment]
Hello @nk521, I think only very few improvements are remaining for this PR to be merged.
Let us know if you do not plan to contribute to this PR anymore, and we will let someone else take over your work.
Thanks!

**Comment by nk521** [comment]
Hey,

Because of Covid-19 and other circumstances I was not able to complete this PR. And I don't plan to complete this as I won't be returning to my work machine anytime soon.

**Comment by xarkes** [comment]
Alright thanks for the feedback, take care!

**Comment by xarkes** [comment]
Here is a refactoring proposal : https://github.com/radareorg/cutter/tree/pr/2100
I'm closing this PR, following the conversation in new PR.
